Serialize NamedDataStoreOutput into PTD.

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #9125

Update PTD serialization to account for blobs from the NamedDataStoreOutput.

Something we can do in the future is to consolidate tensors (that go through the emitter) and blobs (that come from the NamedDataStore).

Differential Revision: [D70939807](https://our.internmc.facebook.com/intern/diff/D70939807/)

lucylq committed:
Serialize NamedDataStoreOutput into PTD.

Update PTD serialization to account for blobs from the NamedDataStoreOutput.

Something we can do in the future is to consolidate tensors (that go through the emitter) and blobs (that come from the NamedDataStore).

Differential Revision: [D70939807](https://our.internmc.facebook.com/intern/diff/D70939807/)

[ghstack-poisoned]

JacobSzwejbka commented on exir/_serialize/_serialize.py:
```diff
@@ -74,39 +75,54 @@ def serialize_for_executorch(
                         tensor.extra_tensor_info.fully_qualified_name
                     ] = TensorLayout(tensor.scalar_type, tensor.sizes, tensor.dim_order)
 
+    if len(fqn_to_tensor_layout) == 0 and (
```
is it possible for only one of these to be true?

lucylq replied:
Yes:

fqn_to_tensor_layout is for tensors being saved to a separate file,
named_data.external_data is for blobs being saved to a separate file

For training, named_data.external_data should be empty.
For delegate weight sharing, fqn_to_tensor_layout would be empty.

JacobSzwejbka commented on exir/_serialize/_serialize.py:
```diff
@@ -74,39 +75,54 @@ def serialize_for_executorch(
+        return pte, ptd_files
+
+    all_external_files: Set[str] = set()
+    if named_data is not None and len(named_data.external_data) > 0:
```
similar question here, can named_data be not none but the length of the data is > 0 or vice versa?

lucylq replied:
named_data stores blobs that are either saved in the PTE file or saved externally. 
If all named data is marked to be saved in the PTE file, named_data.pte_data is non-empty, and named_data.external_data is empty

JacobSzwejbka commented on extension/flat_tensor/serialize/serialize.py:
```diff
@@ -202,6 +205,24 @@ def to_bytes(self) -> bytes:
+
+    def __init__(self, data: Cord, alignment: Optional[int] = None) -> None:
+        self.data = data
+        self.alignment = alignment or 1
```
TIL you can do this

JacobSzwejbka commented on extension/flat_tensor/serialize/serialize.py:
```diff
@@ -305,19 +360,18 @@ def serialize(
-                    offset=aligned_size(prev_end, self.config.segment_alignment),
-                    size=len(segment),
+                    offset=aligned_size(prev_end, alignment),
+                    size=len(segment.data),
```
what else is in the segment now besides data? Alignment?

JacobSzwejbka replied:
oh i meant you have segment.data. What else is in that object besides the data.

lucylq committed:
Update on "Serialize NamedDataStoreOutput into PTD."


Update PTD serialization to account for blobs from the NamedDataStoreOutput.

Something we can do in the future is to consolidate tensors (that go through the emitter) and blobs (that come from the NamedDataStore).

Differential Revision: [D70939807](https://our.internmc.facebook.com/intern/diff/D70939807/)

[ghstack-poisoned]

lucylq committed:
Update on "Serialize NamedDataStoreOutput into PTD."


Update PTD serialization to account for blobs from the NamedDataStoreOutput.

Something we can do in the future is to consolidate tensors (that go through the emitter) and blobs (that come from the NamedDataStore).

Differential Revision: [D70939807](https://our.internmc.facebook.com/intern/diff/D70939807/)

[ghstack-poisoned]
